src/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs

```diff
@@ -73,7 +76,7 @@ public void ResetMemory()
             {
                 owner.Dispose();
             }
-            else
+            else if (_memoryOwner != null)
```

Comment:
This is fine, but you might also consider writing it as:
```C#
else if (_memoryOwner is byte[] array)
{
    ArrayPool<byte>.Shared.Return(array);
}
```
Either seems fine.

Reply:
I like this, I'll do that.